src/java/org/apache/cassandra/utils/memory/MemtableCleanerThread.java

```diff
@@ -27,61 +33,118 @@
  */
 public class MemtableCleanerThread<P extends MemtablePool> extends InfiniteLoopExecutor
 {
-    private static class Clean<P extends MemtablePool> implements InterruptibleRunnable
+    private static final Logger logger = LoggerFactory.getLogger(MemtableCleanerThread.class);
+
+    public static class Clean<P extends MemtablePool> implements InterruptibleRunnable
     {
         /** The pool we're cleaning */
         final P pool;
 
         /** should ensure that at least some memory has been marked reclaiming after completion */
-        final Runnable cleaner;
+        final MemtableCleaner cleaner;
 
         /** signalled whenever needsCleaning() may return true */
         final WaitQueue wait = new WaitQueue();
 
-        private Clean(P pool, Runnable cleaner)
+        /** This is incremented when a cleaner is invoked and decremented when a cleaner has completed */
+        final AtomicInteger numPendingTasks = new AtomicInteger(0);
+
+        /** The maximum number of pending cleaning tasks, beyond this limit we pass the cleaner the threshold so that the cleaner
+         * will not clean unless it has more than the threshold to clean. */
+        final int maxPendingTasks;
+
+        /** The cleaning threshold, when we have more pending tasks than {@link this#maxPendingTasks}, then we pass this
+         * threshold to the cleaner as a minimum memory to clean *.
+         */
+        final double cleanThreshold;
```

Comment:
If we keep this, I'd perhaps stay w/ `cleanupThreshold` to stay consistent w/ the naming elsewhere. Same goes for several other places where we use `cleanThreshold` currently.